pkg/job/scraper_test.go

```diff
@@ -0,0 +1,454 @@
+				}}, nil
+			},
+			cloudwatchRunFunc: func(_ context.Context, job cloudwatchrunner.Job) ([]*model.CloudwatchData, error) {
+				if job.Namespace() == "custom-namespace" {
+					return nil, errors.New("I failed you")
+				}
+				return []*model.CloudwatchData{
+					{
+						MetricName:          "metric-1",
+						ResourceName:        "resource-1",
+						Namespace:           "aws-namespace",
+						Tags:                []model.Tag{{Key: "tag1", Value: "value1"}},
+						Dimensions:          []model.Dimension{{Name: "dimension1", Value: "value1"}},
+						GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Maximum", Datapoint: aws.Float64(1.0), Timestamp: time.Time{}},
+					},
+				}, nil
+			},
+			expectedResources: []model.TaggedResourceResult{
+				{
+					Context: &model.ScrapeContext{Region: "us-east-1", AccountID: "aws-account-1"},
+					Data: []*model.TaggedResource{
+						{ARN: "resource-1", Namespace: "aws-namespace", Region: "us-east-1", Tags: []model.Tag{{Key: "tag1", Value: "value1"}}},
+					},
+				},
+			},
+			expectedMetrics: []model.CloudwatchMetricResult{
+				{
+					Context: &model.ScrapeContext{Region: "us-east-1", AccountID: "aws-account-1"},
+					Data: []*model.CloudwatchData{
+						{
+							MetricName:          "metric-1",
+							ResourceName:        "resource-1",
+							Namespace:           "aws-namespace",
+							Tags:                []model.Tag{{Key: "tag1", Value: "value1"}},
+							Dimensions:          []model.Dimension{{Name: "dimension1", Value: "value1"}},
+							GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Maximum", Datapoint: aws.Float64(1.0), Timestamp: time.Time{}},
+						},
+					},
+				},
+			},
+			expectedErrs: []job.Error{
+				{AccountID: "aws-account-1", Namespace: "custom-namespace", Region: "us-east-2", RoleARN: "aws-arn-2"},
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			rf := testRunnerFactory{
+				GetAccountFunc:    tc.getAccountFunc,
+				MetadataRunFunc:   tc.metadataRunFunc,
+				CloudwatchRunFunc: tc.cloudwatchRunFunc,
+			}
+			sr := job.NewScraper(logging.NewLogger("", true), tc.jobsCfg, &rf)
+			resources, metrics, errs := sr.Scrape(context.Background())
+
+			changelog, err := diff.Diff(tc.expectedResources, resources)
+			assert.NoError(t, err, "failed to diff resources")
+			assert.Len(t, changelog, 0, changelog)
+
+			changelog, err = diff.Diff(tc.expectedMetrics, metrics)
+			assert.NoError(t, err, "failed to diff metrics")
+			assert.Len(t, changelog, 0, changelog)
+
+			// We don't need the check the exact error or message
+			changelog, err = diff.Diff(tc.expectedErrs, errs, diff.Filter(func(_ []string, _ reflect.Type, field reflect.StructField) bool {
+				return !(field.Name == "Err" || field.Name == "Message")
+			}))
+			assert.NoError(t, err, "failed to diff errs")
+			assert.Len(t, changelog, 0, changelog)
```

Comment:
This uses a struct diff library because of the complexity of the struct being compared. The library can diff two complex structs and by default will ignore slice order on any fields which are a slice type which drastically simplifies the assertions here.